### ISSUE TITLE
Support for nD Vectors 

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#262]
## 🚀 Feature
For the vectors layer we want to support a fully nD model with projections occurring over non-displayed dimensions. For this we will switch to an internal representation of an N x 2 x D array which displays N vectors with start and end point in D dimensions. This representation could be passed to the `Shapes` layer, but within the `Vectors` layer special computations can be used to accelerate display as the type of data is know.

The `Vectors` layer will accept either an N x 2 x D array or an (N1 x N2 ... x ND) x D array where each "pixel" contains a length D vector with the D components of the projection of a line onto the D axes. In the later case (N1 x N2 ... x ND) vectors will be drawn.

Does this sound reasonable @bryantChhun?

**Comment by bryantChhun** [comment]
I'm a little confused about the difference between your proposed "internal representation" and what the layer accepts.

The internal representation is currently (N, 2) coordinates, paired such that consecutive pairs represent line endpoints in (x, y).  This is largely a layover from the GL lines approach and can absolutely be adjusted in any way optimal for mesh rendering.  And, It sounds like the N x 2 x D enables the data to pass to `Shapes` layer.  OK.

I'm confused about the dimensionality here, please bear with me.  As I understand it :
N x 2 x D means: 
N = number of vectors
2 = coordinate in dimension... x, y? that represents an end point?
D = number of dimensions?
Shouldn't it be N x D x D, so that a scalar is defined for each of D dimensions, which defines a coordinate in D dimensions?  What does the final "x D" mean then?

Additionally there is the topic of what the `Vectors` layer accepts from the user (which is or is not different from the "internal" representation):  I interpret the N x 2 x D as reflecting the "coordinate" type of data input.  And the (N1 x N2 x ND) x D as reflecting the "image" type of data input.  But, as it is currently implemented, the "coordinate" type should contain **both coordinates and projections**, which means a minimum of N x 4 x D.  In this case N represents a true vector, since it encodes both position and direction.  Internally, an N x 4 input is transformed into a 2N x 2 list of coordinates.

So the next question is: What format, exactly, does the user supply for "coordinate" type vectors?

**Comment by sofroniewn** [comment]
Sorry I should have been more clear - here is what I had in mind

`N x 2 x D means`:
- N = number of vectors
- 2 = start point and end point
- D = number of dimensions e.g. (z, y, x)

For "internal representation" we'd use `N x 2 x D`, but the layer accepts both 
- `N x 2 x D` which is "coordinate" type data
- `(N1 x N2 x ND) x D` which is "image" type data where each pixel is a length `D` vector of the projections

Does this make it more clear? I think the confusion was around the `2` being used to represent the start-point and end-point and not the `(x, y)` coordinates. Hopefully this addresses your comments about encoding both position and direction

**Comment by sofroniewn** [comment]
One thing I want to be more explicit about before starting implementation work - is what should the behavior be of vectors that slice across planes? This has been discuss a little in the context of `nD` shapes in #296 

Say we have two vectors in a 3D space with the following start and end points
```python
[[[0, y0, x0], [0, y1, x1]],
 [[1, y2, x2],  [1, y3, x3]]]
```
where `[y0, x0]` and `[y1, x1]` are the start and end points of the vector in the zeroth z-slice etc. In this case it's easy, there should just be two vectors drawn, one in the zeroth z-slice and one in the first z-slice.

What happens though if someone tries to pass
```python
[[[0, y0, x0], [1, y1, x1]]]
```
Our options are as follows:
1. For the `nD` shapes layer we ignore all such shapes, and so for consistency we could do that again here.
2. Or we could take a max projection over the z-range, and draw vectors from `[y0, x0]` to `[y1, x1]` in both the zeroth z-slice and one in the first z-slice, but no subsequent z-slices.
3. Or we could do some weird slicing and draw a vector from `[y0, x0]` to `([y0, x0]+[y1, x1])/2` in the zeroth z-slice, a vector from `([y0, x0]+[y1, x1])/2` to `[y1, x1]` in the first z-slice, and none in subsequent z-slices.

I worry that the logic in (3) starts getting a little intense, and will not feel intuitive, but just be confusing. I'm not quite sure about (2) either. (1) is simple to implement, is consistent with the shapes layer, and at least provides us with some `nD` support.

How do @bryantChhun, @jni, and @royerloic feel about this?

**Comment by bryantChhun** [comment]
Putting some thought into this ...  What Shalin and I both believe is proper is to draw the intersection of a plane with a line, which results in a point (or segment) that travels along x-y as you move through z.  If our "plane" is really a "sheet" with thickness, then this means the vector isn't a point, but a small segment of the full-length vector.  Unfortunately, this is most analgous to #3 above.

Is there some library that could help us calculate these?  Wouldn't this geometry-intersection approach be a normal thing to do with shaders, rather than algebraically like above?

I'm also ok with going route #1 above and shelving this for a later time...

**Comment by sofroniewn** [comment]
@bryantChhun, yes this is like (3) above. We'd have to assume that all planes are evenly spaced apart, but it should be possible. The calculations won't be too bad.

An alternative suggested by @jni is just to show the whole vector in the plane of the starting location.

I propose that for we initially make a PR implementing (1) (i.e. only showing vectors entirely in a plane and ignoring / warning about the others) and then make a follow up PR implementing something more advanced afterwards. Possibly as we're implementing and reviewing (1) a solution will become clear.

If that sounds good, I'm happy to start working on the new PR

**Comment by bryantChhun** [comment]
Yup sounds good.  Thanks for working on this PR.

**Comment by jni** [comment]
I can't disagree more with the Shalin's and @bryantChhun's assessment... Please correct me if I'm wrong, but drawing a vector (1, 1, 3) at position (4, 5, 6) means that the fluid/particle at position (4, 5, 6) has velocity (1, 1, 3). You visualise this by placing a vector (1, 1, 3) either starting at (most commonly) or centered at (haven't seen this but would make some sense) (4, 5, 6). Seeing a fragment of a vector at (5, 6, 9) makes no sense to me.

Obviously I'm happy with your approach @sofroniewn :joy:

Note: I consider this different from something like `lines_nd`, which indicates *linkage* between two points. For that, I absolutely would support the "lineslice" visualisation model. But I don't think such a layer should be called "vectors".

**Comment by bryantChhun** [comment]
> I can't disagree more with the Shalin's and @bryantChhun's assessment... Please correct me if I'm wrong, but drawing a vector (1, 1, 3) at position (4, 5, 6) means that the fluid/particle at position (4, 5, 6) has velocity (1, 1, 3). You visualise this by placing a vector (1, 1, 3) either starting at (most commonly) or centered at (haven't seen this but would make some sense) (4, 5, 6). Seeing a fragment of a vector at (5, 6, 9) makes no sense to me.
> 
> Obviously I'm happy with your approach @sofroniewn 😂
> 
> Note: I consider this different from something like `lines_nd`, which indicates _linkage_ between two points. For that, I absolutely would support the "lineslice" visualisation model. But I don't think such a layer should be called "vectors".

OK I think I've come around on this.  It is important to distinguish path vs vectors, as you describe in your note.  Here we definitely want to describe a vector that encodes both orientation and magnitude.  However, because viewing a (1, 1, 3) vector starting at (4, 5, 6) wholly in plane (_, _, 6) is a projection, we have to somehow convey when a vector is fully in-plane or a projection.  For this, I think your color scheme could work (to be implemented in the future?).
